src/libcore/num/dec2flt/rawfp.rs

```diff
@@ -296,14 +285,14 @@ pub fn encode_normal<T: RawFloat>(x: Unpacked) -> T {
         "encode_normal: exponent out of range");
     // Leave sign bit at 0 ("+"), our numbers are all positive
     let bits = (k_enc as u64) << T::EXPLICIT_SIG_BITS | sig_enc;
-    T::from_bits(bits)
+    T::from_bits(bits.try_into().unwrap_or_else(|_| unreachable!()))
 }
 
 /// Construct a subnormal. A mantissa of 0 is allowed and constructs zero.
 pub fn encode_subnormal<T: RawFloat>(significand: u64) -> T {
     assert!(significand < T::MIN_SIG, "encode_subnormal: not actually subnormal");
     // Encoded exponent is 0, the sign bit is 0, so we just have to reinterpret the bits.
-    T::from_bits(significand)
+    T::from_bits(significand.try_into().unwrap_or_else(|_| unreachable!()))
```

Comment:
This seems to be a step back, isn't there a way to keep this using just from_bits? Do you know if it has any performance impact?

The assertion `assert!(bits < u32::MAX as u64, "f32::from_bits: too many bits");` from the old code has been lost, but I suppose it is unreachable here.

Reply:
I wouldn't worry about performance impact, this is only called near the end of the slowest of slow paths.

It's still ugly as hell but that ugliness will have to be somewhere unless all the dec2flt code is restructured to work with the `Bits` associated type instead of u64 (which would probably not be pretty either). Maybe it could move into `RawFloat::from_bits`? (But if so, it should gain a nicer panic message than `unreachable`)

Reply:
I'll try seeing what needs to be done to use `Bits` instead of `u64` here.

Reply:
That's probably far too big a refactor to be worthwhile for this PR.

Reply:
Moving the try_from into from_bits seems more fruitful.